src/measure_theory/kernel/basic.lean

```diff
@@ -0,0 +1,295 @@
+end
+
+end sum
+
+section s_finite
+
+/-- A kernel is s-finite if it can be written as the sum of countably many finite kernels. -/
+class is_s_finite_kernel (κ : kernel α β) : Prop :=
+(tsum_finite : ∃ κs : ℕ → kernel α β, (∀ n, is_finite_kernel (κs n)) ∧ κ = kernel.sum κs)
+
+@[priority 100]
+instance is_finite_kernel.is_s_finite_kernel [h : is_finite_kernel κ] : is_s_finite_kernel κ :=
+⟨⟨λ n, if n = 0 then κ else 0,
+  λ n, by { split_ifs, exact h, apply_instance, },
+  begin
+    ext a s hs,
+    rw kernel.sum_apply' _ _ hs,
+    have : (λ i, ((ite (i = 0) κ 0) a) s) = λ i, ite (i = 0) (κ a s) 0,
+    { ext1 i, split_ifs; refl, },
+    rw [this, tsum_ite_eq],
+  end⟩⟩
+
+/-- A sequence of finite kernels such that `κ = kernel.sum (seq κ)`. See `is_finite_kernel_seq`
+and `kernel_sum_seq`. -/
+noncomputable
+def seq (κ : kernel α β) [h : is_s_finite_kernel κ] :
+  ℕ → kernel α β :=
+h.tsum_finite.some
+
+lemma kernel_sum_seq (κ : kernel α β) [h : is_s_finite_kernel κ] :
+  kernel.sum (seq κ) = κ :=
+h.tsum_finite.some_spec.2.symm
+
+lemma measure_sum_seq (κ : kernel α β) [h : is_s_finite_kernel κ] (a : α) :
+  measure.sum (λ n, seq κ n a) = κ a :=
+by rw [← kernel.sum_apply, kernel_sum_seq κ]
+
+instance is_finite_kernel_seq (κ : kernel α β) [h : is_s_finite_kernel κ] (n : ℕ) :
+  is_finite_kernel (kernel.seq κ n) :=
+h.tsum_finite.some_spec.1 n
+
+instance is_s_finite_kernel.add (κ η : kernel α β) [is_s_finite_kernel κ] [is_s_finite_kernel η] :
+  is_s_finite_kernel (κ + η) :=
+begin
+  refine ⟨⟨λ n, seq κ n + seq η n, λ n, infer_instance, _⟩⟩,
+  rw [sum_add, kernel_sum_seq κ, kernel_sum_seq η],
+end
+
+lemma is_s_finite_kernel.finset_sum {κs : ι → kernel α β} (I : finset ι)
+  (h : ∀ i ∈ I, is_s_finite_kernel (κs i)) :
+  is_s_finite_kernel (∑ i in I, κs i) :=
+begin
+  classical,
+  unfreezingI
+  { induction I using finset.induction with i I hi_nmem_I h_ind h,
+    { rw [finset.sum_empty], apply_instance, },
+    { rw finset.sum_insert hi_nmem_I,
+      haveI : is_s_finite_kernel (κs i) := h i (finset.mem_insert_self _ _),
+      haveI : is_s_finite_kernel (∑ (x : ι) in I, κs x),
+        from h_ind (λ i hiI, h i (finset.mem_insert_of_mem hiI)),
+      exact is_s_finite_kernel.add _ _, }, },
+end
+
+lemma is_s_finite_kernel_sum_of_denumerable [denumerable ι] {κs : ι → kernel α β}
+  (hκs : ∀ n, is_s_finite_kernel (κs n)) :
+  is_s_finite_kernel (kernel.sum κs) :=
+begin
+  let e : ℕ ≃ (ι × ℕ) := denumerable.equiv₂ ℕ (ι × ℕ),
+  refine ⟨⟨λ n, seq (κs (e n).1) (e n).2, infer_instance, _⟩⟩,
+  have hκ_eq : kernel.sum κs = kernel.sum (λ n, kernel.sum (seq (κs n))),
+  { simp_rw kernel_sum_seq, },
+  ext1 a,
+  ext1 s hs,
+  rw hκ_eq,
+  simp_rw kernel.sum_apply' _ _ hs,
+  change ∑' i m, seq (κs i) m a s = ∑' n, (λ im : ι × ℕ, seq (κs im.fst) im.snd a s) (e n),
+  rw e.tsum_eq,
+  { rw tsum_prod' ennreal.summable (λ _, ennreal.summable), },
+  { apply_instance, },
+end
+
+lemma is_s_finite_kernel_sum [countable ι] {κs : ι → kernel α β}
```

Comment:
Should this be an instance? Same question for a few lemmas above.

Reply:
I wrote them as lemmas because I can't get them to work as instances. They are not found.